dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ExceptionHelper.java

```diff
@@ -0,0 +1,153 @@
+    @SuppressWarnings("checkstyle:IllegalType")
+    private static List<Throwable> listThrowables(Throwable throwable) {
+        final LinkedHashSet<Throwable> list = new LinkedHashSet<>();
+        while (throwable != null && !list.contains(throwable)) {
+            list.add(throwable);
+            throwable = throwable.getCause();
+        }
+        return new ArrayList<>(list);
+    }
+
+    /**
+     * Calculates useful information about the exception/throwable.
+     */
+    public Info info() {
+        return handleJavaThrowable()
+            .or(() -> handleConstraintViolationException())
+            .or(() -> handlePersistenceException())
+            .or(() -> result(HttpStatus.SC_BAD_REQUEST))
+            .get();
+    }
+
+    /**
+     * Calculates a message that describes the exception/throwable.
+     */
+    public String message() {
+        return info().message();
+    }
+
+    /**
+     * Calculates a suggested HTTP status code for the exception/throwable.
+     */
+    public int status() {
+        return info().status();
+    }
+
+    private <T extends Throwable> Optional<T> findThrowable(Class<T> klass) {
+        return throwables.stream().filter(klass::isInstance).map(klass::cast).findFirst();
+    }
+
+    private <T extends Throwable> boolean hasThrowable(Class<T> klass) {
+        return findThrowable(klass).isPresent();
+    }
+
+    private Optional<Info> handleJavaThrowable() {
+        String className = throwable.getClass().getName();
+        return (className.startsWith("java.") || className.startsWith("javax."))
+            ? result(HttpStatus.SC_INTERNAL_SERVER_ERROR)
+            : NONE;
+    }
+
+    private Optional<Info> handleConstraintViolationException() {
+        return findThrowable(ConstraintViolationException.class).flatMap(this::mapConstraintViolationException);
+    }
+
+    private Optional<Info> mapConstraintViolationException(ConstraintViolationException c) {
+        String message = mapConstraintName(c.getConstraintName());
+        return result(message, HttpStatus.SC_CONFLICT);
+    }
+
+    @SuppressWarnings("checkstyle:indentation")
+    private String mapConstraintName(String name) {
+        return name == null
+            ? "a database constraint was violated"
+            : switch (name.toLowerCase()) {
+                case "aliases_are_unique" -> same("an entry", "alias");
+                case "case_insensitive_apptool_toolname" -> same("a tool", "name");
+                case "case_insensitive_notebook_workflowname" -> same("a notebook", "name");
+                case "case_insensitive_toolname" -> same("a tool", "name");
+                case "case_insensitive_workflow_workflowname" -> same("a workflow", "name");
+                case "check_valid_doi" -> "the DOI is not valid";
+                case "check_valid_orcid" -> "the ORCID is not valid";
```

Comment:
Although it sounds redundant, with "ID" already being part of "ORCID", you should say "ORCID identifier" or "ORCID iD" (yuck on "iD", but that's what they say to do).

https://info.orcid.org/brand-guidelines/